File: Codes/0063-unique-paths-ii_20250208.cpp
/*
 * @Author: LetMeFly
 * @Date: 2025-02-08 09:36:16
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2025-02-08 09:53:50
 */
#ifdef _WIN32
#include "_[1,2]toVector.h"
#endif

class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        for (int i = 0; i < obstacleGrid.size(); i++) {
            for (int j = 0; j < obstacleGrid[0].size(); j++) {
                if (i == 0 && j == 0) {
                    obstacleGrid[i][j] = obstacleGrid[i][j] ? 0 : 1;
                    continue;
                } else if (obstacleGrid[i][j]) {
                    obstacleGrid[i][j] = -1;
                    continue;
                }
                if (i > 0 && obstacleGrid[i - 1][j] != -1) {
                    obstacleGrid[i][j] += obstacleGrid[i - 1][j];
                }
                if (j > 0 && obstacleGrid[i][j - 1] != -1) {
                    obstacleGrid[i][j] += obstacleGrid[i][j - 1];
                }
            }
        }
        return max(obstacleGrid.back().back(), 0);
    }
};

#ifdef _WIN32
/*
[[0,0]]

1
*/
int main() {
    string s;
    Solution sol;
    while (cin >> s) {
        vector<vector<int>> v = stringToVectorVector(s);
        cout << sol.uniquePathsWithObstacles(v) << endl;
    }
    return 0;
}
#endif

File: Codes/0063-unique-paths-ii_20250208.java
/*
 * @Author: LetMeFly
 * @Date: 2025-02-08 09:55:20
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2025-02-08 10:02:07
 */
class Solution {
    public int uniquePathsWithObstacles(int[][] a) {
        if (a[0][0] == 1) {
            return 0;
        }
        a[0][0] = 1;
        for (int i = 0; i < a.length; i++) {
            for (int j = 0; j < a[0].length; j++) {
                if (a[i][j] == 1 && i + j > 0) {
                    a[i][j] = 0;
                    continue;
                }
                if (i > 0) {
                    a[i][j] += a[i - 1][j];
                }
                if (j > 0) {
                    a[i][j] += a[i][j - 1];
                }
            }
        }
        return a[a.length - 1][a[0].length - 1];
    }
}